verilog/analysis/checkers/always_ff_only_local_blocking_rule.cc

```diff
@@ -0,0 +1,167 @@
+
+absl::string_view AlwaysFFOnlyLocalBlockingRule::Name() {
+  return "always-ff-only-local-blocking";
+}
+const char AlwaysFFOnlyLocalBlockingRule::kTopic[] = "sequential-logic";
+const char AlwaysFFOnlyLocalBlockingRule::kMessage[] =
+    "Use blocking assignments only for locals inside \'always_ff\' sequential blocks.";
+
+std::string AlwaysFFOnlyLocalBlockingRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat("Checks that there are no occurrences of ",
+                      "blocking assignment to non-locals in sequential logic.");
+}
+
+
+static bool const  DEBUG = false;
+
+void AlwaysFFOnlyLocalBlockingRule::HandleSymbol(verible::Symbol const& symbol,
+                                           SyntaxTreeContext const& context) {
+
+  static Matcher const  always_ff_matcher = NodekAlwaysStatement(AlwaysFFKeyword());
+  static Matcher const  block_matcher = NodekBlockItemStatementList();
+  static Matcher const  decl_matcher = NodekDataDeclaration();
+  static Matcher const  var_matcher = NodekRegisterVariable();
+  static Matcher const  asgn_blocking_matcher = NodekNetVariableAssignment();
+  static Matcher const  asgn_modify_matcher = NodekAssignModifyStatement();
+  static Matcher const  asgn_incdec_matcher = NodekIncrementDecrementExpression();
+  static Matcher const  ident_matcher = NodekUnqualifiedId();
+
+  // Determine depth in syntax tree and discard state from branches already left
+  int const  depth = context.size();
+  if(depth <= this->inside)  this->inside = 0;
+  while(depth <= scopes.top().first) {
+    scopes.pop();
+    if(DEBUG) {
+      std::cerr
+          << "POPped to scope DEPTH=" << scopes.top().first
+          << "; #LOCALS=" << scopes.top().second << std::endl;
+    }
+  }
+  locals.resize(scopes.top().second);
+
+  verible::matcher::BoundSymbolManager  symbol_man;
+
+  // Check for entering an always_ff block
+  if(always_ff_matcher.Matches(symbol, &symbol_man)) {
+    if(DEBUG)  std::cerr << "always_ff @DEPTH=" << depth << std::endl;
+    this->inside = depth;
+  }
+  else if(this->inside) {
+    // Open up begin-end block
+    if(block_matcher.Matches(symbol, &symbol_man)) {
+      if(DEBUG) {
+        std::cerr
+          << "PUSHing scope: DEPTH=" << depth
+          << "; #LOCALs inherited=" << locals.size() << std::endl;
+      }
+      scopes.emplace(depth, locals.size());
+    }
+    // Collect local variable declarations
+    else if(decl_matcher.Matches(symbol, &symbol_man)) {
+      auto& cnt = scopes.top().second;
+      for(auto const& var : SearchSyntaxTree(symbol, var_matcher)) {
+        if(auto const *const  node = dynamic_cast<verible::SyntaxTreeNode const*>(var.match)) {
+          if(auto const *const  ident = dynamic_cast<verible::SyntaxTreeLeaf const*>(node->children()[0].get())) {
+            absl::string_view const  name = ident->get().text();
+            if(DEBUG)  std::cerr << "Registering '" << name << '\'' << std::endl;
```

Comment:
For these DEBUG outputs, I recommend using VLOG(), which is the common way to create debug messages in this codebase. Check out other places in the code-base for examples.

Here, you could write
```c++
  VLOG(1) << "Registering '" << name << "'";
```
(It will take care of adding a newline automatically).

To output these messages, you can enable the logging, e.g. by setting an environment variable:

```
GLOG_v=1 bazel-bin/verilog/analysis/checkers/always_ff_only_local_blocking_rule_test
```

Reply:
ACK. Will do.